### PR TITLE
Fix MessageDeduplication#inactiveProducers may not be persistence correctly

### PR DESCRIPTION
**Pull request by shibd**
### Motivation

#15204 

In the current implementation, When the first time execute `purgeInactiveProducers`, Although the produces does not expire, it removed directly from the collection(464 line). The will result in these producers never being remove.

https://github.com/apache/pulsar/blob/9861dfb1208c4b6b8a1f17ef026e9af71c3e784c/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java#L454-L472

### Modifications

1. It is removed from the collection only when the producer is inactive.
2. Take a snapshot after each removal of an inactive producer. When `managedLedger.getLastConfirmedEntry` equals `managedCursor.getMarkDeletedPosition()`, The`deduplication-snapshot-monitor` thread does not trigger a snapshot. The persistence these producers only the next time a message is produced, The can be confusing for users.

```
        PositionImpl position = (PositionImpl) managedLedger.getLastConfirmedEntry();
        if (position == null) {
            return;
        }
        PositionImpl markDeletedPosition = (PositionImpl) managedCursor.getMarkDeletedPosition();
        if (markDeletedPosition != null && position.compareTo(markDeletedPosition) <= 0) {
            return;
        }
```

### Documentation
- [x] `no-need-doc`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java [hunk 1]
```diff
@@ -456,19 +457,22 @@ public synchronized void purgeInactiveProducers() {
                 .toMillis(pulsar.getConfiguration().getBrokerDeduplicationProducerInactivityTimeoutMinutes());
 
         Iterator<java.util.Map.Entry<String, Long>> mapIterator = inactiveProducers.entrySet().iterator();
+        boolean hasInactive = false;
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java:460]: At glance, Looks like this value is always false.

Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java:460]: I think I can do a more deep review later, thanks.

Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java:460]: Thanks, you are right. I missed a line. Fixed, PTAL.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java [hunk 2]
```diff
@@ -456,19 +457,23 @@ public synchronized void purgeInactiveProducers() {
...
             if (lastActiveTimestamp < minimumActiveTimestamp) {
                 log.info("[{}] Purging dedup information for producer {}", topic.getName(), producerName);
+                mapIterator.remove();
                 highestSequencedPushed.remove(producerName);
                 highestSequencedPersisted.remove(producerName);
+                hasInactive = true;
             }
         }
+        if (hasInactive) {
+            takeSnapshot(getManagedCursor().getMarkDeletedPosition());
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java:475]: If there are a large number of topics, snapshots may be created concurrently. We should avoid creating snapshots frequently. This is now done by the timer thread.

Comment (315157973) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java:475]: `checkMessageDeduplicationInfo` is not a frequent operation, but it traverses all topics each time when it is executed. If many topics have Producers that need to be purged, a large number of snapshots will be generated.
Currently, `takeSnapshot` is automatically executed every 120 seconds. Therefore, we do not need to generate a snapshot after each topic is purged.

Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java:475]: Thanks your reply, the main reason I triggered a snapshot in the `purgeInactiveProducers` method was. If we rely on `deduplication-snapshot-monitor` timer to take snapshot, that may not be executed. Because we did the optimizations in the `takeSnapshot()` method.
```
        PositionImpl markDeletedPosition = (PositionImpl) managedCursor.getMarkDeletedPosition();
        if (markDeletedPosition != null && position.compareTo(markDeletedPosition) <= 0) {
            return;
        }
```

When `dup.cursor.markDeletedPosition == ledger.lastConfirmedEntry` and all producers stoped. In this case, we will not be able to clean up these producers. Unless new data is written and take a snapshot. The can be confusing for users.

> but it traverses all topics each time when it is executed. If many topics have Producers that need to be purged, a large number of snapshots will be generated.

In effect, automatically `takeSnapshot` also traverses all topics.

> we do not need to generate a snapshot after each topic is purged. Or, wait until all topics are cleaned up and create a snapshot.

This PR is only take a snapshot when there are inactive topics

In conclusion, the snapshot here is to solve the problem described above, `checkMessageDeduplicationInfo` is not a frequent operation.

Thanks again, If you have any suggestions, please continue the discussion.

**Review by 315157973 — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Review by Jason918 — APPROVED** [review]
LGTM

**Review by congbobo184 — APPROVED** [review]

**Review by mattisonchao — APPROVED** [review]

**Review by nodece — APPROVED** [review]

**Review by RobertIndie — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/MessageDuplicationTest.java [hunk 3]
```diff
@@ -187,26 +190,28 @@ public void testInactiveProducerRemove() throws Exception {
...
+        // Modify the inactive time of produce2 and produce3
         // messageDeduplication.purgeInactiveProducers() will remove producer2 and producer3
-        map.put(producerName2, System.currentTimeMillis() - 70000);
-        map.put(producerName3, System.currentTimeMillis() - 70000);
+        inactiveProducers.put(producerName2, System.currentTimeMillis() - 70000);
+        inactiveProducers.put(producerName3, System.currentTimeMillis() - 70000);
+        // Try second purgeInactive, produce2 and produce3 is inactive.
         messageDeduplication.purgeInactiveProducers();
-        assertFalse(map.containsKey(producerName2));
-        assertFalse(map.containsKey(producerName3));
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/MessageDuplicationTest.java:200]: Why are you removing this assertion?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/MessageDuplicationTest.java (hunk 3 above)
Comment (shibd) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/MessageDuplicationTest.java:200]: Thank you for your review, this assertion can be kept, I reverted it.

Comment (shibd) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/MessageDuplicationTest.java:200]: @eolivelli Can you help me again take a look?

**Comment by codelipenghui** [comment]
@eolivelli Please help review this PR, thanks.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/MessageDuplicationTest.java (hunk 3 above)
Comment (shibd) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/persistent/MessageDuplicationTest.java:200]: ping @eolivelli

**Review by eolivelli — APPROVED** [review]
Lgtm
